File: solutions/lua/prism/1/prism.lua
local function find_sequence(start, prisms)
  local sequence = {}
  local x, y, angle = start.x, start.y, start.angle % 360

  while true do
    local state = {
      distance = math.huge,
      id = nil,
      x = nil,
      y = nil,
      angle = nil
    }

    for _, prism in ipairs(prisms) do
      local dx = prism.x - x
      local dy = prism.y - y

      if dx ~= 0 or dy ~= 0 then
        local angle_to_prism = math.deg(math.atan(dy, dx)) % 360

        if math.min(math.abs(angle_to_prism - angle), math.abs(angle_to_prism - angle + 360)) < 1e-2 then
          local distance = math.sqrt(dx * dx + dy * dy)

          if distance < state.distance then
            state = {
              distance = distance,
              id = prism.id,
              x = prism.x,
              y = prism.y,
              angle = (angle + prism.angle) % 360
            }
          end
        end
      end
    end

    if state.id then
      table.insert(sequence, state.id)
      x, y, angle = state.x, state.y, state.angle
    else
      return sequence
    end
  end
end

return { find_sequence = find_sequence }

File: solutions/go/prism/1/prism.go
package prism

import "math"

type Position struct {
	x     float64
	y     float64
	angle float64
}

type Prism struct {
	id    int
	x     float64
	y     float64
	angle float64
}

func normalizeAngle(a float64) float64 {
	a = math.Mod(a, 360)
	if a < 0 {
		a += 360
	}
	return a
}

func angleDifference(a1, a2 float64) float64 {
	diff := math.Abs(a1 - a2)
	if diff > 180 {
		diff = 360 - diff
	}
	return diff
}

func FindSequence(start Position, prisms []Prism) []int {
	sequence := []int{}
	x, y := start.x, start.y
	angle := normalizeAngle(start.angle)

	for {
		minDistance := math.Inf(1)
		var nextID int
		var nextX, nextY, nextAngle float64
		found := false

		for _, prism := range prisms {
			dx := prism.x - x
			dy := prism.y - y

			if dx != 0 || dy != 0 {
				angleToPrism := normalizeAngle(math.Atan2(dy, dx) * 180 / math.Pi)

				if angleDifference(angleToPrism, angle) < 1e-2 {
					distance := math.Sqrt(dx*dx + dy*dy)
					if distance < minDistance {
						minDistance = distance
						nextID = prism.id
						nextX = prism.x
						nextY = prism.y
						nextAngle = normalizeAngle(angle + prism.angle)
						found = true
					}
				}
			}
		}

		if found {
			sequence = append(sequence, nextID)
			x, y, angle = nextX, nextY, nextAngle
		} else {
			return sequence
		}
	}
}